### DISCUSSION TITLE
Having Trouble With Chinese PDF

### DISCUSSION THREAD

**Discussion by katieyang** [microsoft/AI-Chat-App-Hack#56] [Q&A]
Hello,

We are working hard on the hackathon and have run into a problem where Azure seems to be erroring. We think it's because we fed some Chinese documents in and it is having trouble searching/parsing for those.

This is the error we are seeing:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/15933193/9cda634c-78ce-46fe-bbab-e29cc74a8ffe)

It does seem to be working when it uses the English sources, even when I ask questions in Chinese.

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/15933193/d4b6c354-fb97-4067-9142-017e87eb0bf2)

We are using free tier.

Thanks for your help.

I have attached the one of the pdfs in Chinese in case helpful.
[Jack Ma's Bund Finance Summit Speech 2020 on the financial system and risk.pdf](https://github.com/microsoft/AI-Chat-App-Hack/files/14151802/Jack.Ma.s.Bund.Finance.Summit.Speech.2020.on.the.financial.system.and.risk.pdf)

**Comment by zedhaque** [comment]
Hi
Can you check your .env value for 
AZURE_SEARCH_ANALYZER_NAME
- this needs to be changed to support Chinese language you can find the language code - https://learn.microsoft.com/en-au/dotnet/api/microsoft.azure.search.models.field.analyzer?view=azure-dotnet-legacy&viewFallbackFrom=azure-dotnet

Your current index is probably built using en.microsoft (default English language). You will need to drop the current index from Azure portal and re-run scripts/prepdocs.sh with the Chinese pdf file again. This will then re-create the index with correct Analyzer setting.

You will also need to change your 
AZURE_SEARCH_QUERY_LANGUAGE
So during query it supports Chinese.  Note this will require you to run “azd deploy” again. 

If you have already set above variables correctly. You might want to change the prompt to be Chinese. Are you planning to support both English and Chinese or just one language? 

Hope this helps.

**Reply by katieyang** [reply]
We'll give it a try! We were hoping to support both languages.

**Reply by AshleyZhao** [reply]
If we want to support both languages, is it possible on the free tier?

I saw that we need to implement a language analyzer somewhere, but not sure if that is a valid solution.

**Reply by pamelafox** [reply]
Is it that you have both English and Chinese documents, and you want to support both English and Chinese questions on those documents?

**Reply by katieyang** [reply]
Yes, we have both English and Chinese documents and want to support both English and Chinese prompts.

**Reply by pamelafox** [reply]
@srbalakr Could you advise as to the best analyzer to use for this situation?

**Reply by pamelafox** [reply]
I've talked to search team and they pointed me to this article:
https://learn.microsoft.com/en-us/azure/search/search-language-support

They recommend creating a field for each language in your search index, like content_en and content_zh, and using the appropriate analyzer for each field. You'd need to populate those fields at indexing time depending on the language of the document. You can then search on both those fields. An additional enhancement is to only search specifically the field for the language of the user's query, but you'd need to use ChatGPT to select the right language before doing the query. You could modify search_sources in chatreadretrieveread.py to also accept a language parameter, and use that to decide which fields to search.

That's a lot of work, so you probably just want to do whatever works for now, just wanted to share ideas.

**Reply by katieyang** [reply]
Thanks for all the help Pamela. We have some ideas of where to go with this. Will be working on it more this weekend.

**Comment by pamelafox** [comment]
Hm, I don't think you should see an error if it can't find sources, so it'd be useful to figure out exactly what is causing the OpenAI error. This link has instructions on downloading the logs with the full error traceback:
https://blog.pamelafox.org/2023/06/tips-for-debugging-flask-deployments-to.html#download-logs

**Reply by AshleyZhao** [reply]
Thanks, looks like it's working now.

Just curious, can the solution above be applied to support three languages or more?

**Comment by zedhaque** [comment]
Cool work. Are you using one index or multiple language specific index? Keen to hear how you solved it. 

Yes you can build multi lingual chatbot each backed by different language sources. This requires code modification as per your chatbot requirements and how you plan to roll out your chatbot etc.

**Comment by AshleyZhao** [comment]
We tried implementing the language analyzer (default) but it seems to still not be able to parse Chinese PDFs properly. While examining the logs, we see that there could be a potential error with maximum tokens exceeded (we are using OpenAI):

```
2024-02-04T15:33:16.2382473Z openai.BadRequestError: Error code: 400 - {'error': {'message': "This model's maximum context length is 4097 tokens. However, your messages resulted in 4192 tokens. Please reduce the length of the messages.", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}
```

We are looking for advice on how to decrease the number of tokens in one chunk of text for multiple languages.

(Here's the full log)
[chatbot_error_log.zip](https://github.com/microsoft/AI-Chat-App-Hack/files/14157374/chatbot_error_log.zip)

**Reply by pamelafox** [reply]
Hm, I wonder if there is a bug in our code that tries to estimate the token count for Chinese languages. We've had issues with token counting in CJK languages in the past. Could you possibly share the PDF? You could email to pamelafox @ microsoft . com if so. Otherwise I'll try to find a chinese-language PDF to parse.

**Reply by katieyang** [reply]
See attached two PDFs (I'm not quite sure which one is the one causing the problem). Is there a place we could override the number of tokens it breaks the pdfs down into from how it is determined right now?

[Jack Ma Third Annual Young Entrepreneurs Summit 2020.pdf](https://github.com/microsoft/AI-Chat-App-Hack/files/14158721/Jack.Ma.Third.Annual.Young.Entrepreneurs.Summit.2020.pdf)

[Jack Ma's Bund Finance Summit Speech 2020 on the financial system and risk.pdf](https://github.com/microsoft/AI-Chat-App-Hack/files/14158722/Jack.Ma.s.Bund.Finance.Summit.Speech.2020.on.the.financial.system.and.risk.pdf)

**Reply by pamelafox** [reply]
Check out scripts/prepdocslibs/embeddings.py, split_text_into_batches functions. That's where I'm testing a fix now. Thanks for the PDFs!

**Reply by pamelafox** [reply]
Ah wait, I realize your error is during the chatting stage, not the ingestion phase. That's an issue with the message builder logic then. Let me try to replicate, I was able to ingest the PDFs.

**Reply by pamelafox** [reply]
The logic for building up the messages list is in get_messages_from_history in chatapproach.py. I'm wondering if the chinese text in the sources is requiring so many tokens that it cannot fit even a single question in the request. In that case, you either need to:

1) store shorter chunks in index
2) trim the system prompt
3) use a gpt model with a higher context window

**Reply by pamelafox** [reply]
I haven't been able to replicate fully yet as I need to update my analyzer, but I made a video showing how I'd use the debugger to see how it was counting the tokens:
https://www.youtube.com/watch?v=GIuj_w2HlT0

Does it only happen after you've asked a few questions, or does it happen on first question too?

**Reply by AshleyZhao** [reply]
Thank you,
Yes it happens all the time AFAIK

**Reply by pamelafox** [reply]
Okay, what did you set for AZURE_SEARCH_ANALYZER_NAME and AZURE_SEARCH_QUERY_SPELLER? I want to mimic your environment.

**Reply by katieyang** [reply]
azd env set AZURE_SEARCH_QUERY_LANGUAGE zh-cn
azd env set AZURE_SEARCH_QUERY_SPELLER none
azd env set AZURE_SEARCH_ANALYZER_NAME zh-Hans.lucene

**Reply by pamelafox** [reply]
Great, I replicated the issue. So the problem is that the splitting algorithm in our prepdocs script splits based on character length, not tokens, and 1000 characters in Chinese requires a lot more tokens than 1000 characters in English. The request takes up too much tokens for the sources alone, and doesn't have enough space left for the system message or user question.

I was able to reduce the token count enough by retrieving only 2 results from the search index:

<img width="1431" alt="Screenshot 2024-02-05 at 5 28 47 AM" src="https://github.com/microsoft/AI-Chat-App-Hack/assets/297042/2012cddc-fed0-46cb-b1d8-92e4a5176599">

I think your best bet is to change the chunk sizes in textsplitter.py:

```
        self.max_section_length = 1000
        self.sentence_search_limit = 100
        self.section_overlap = 100
```

I suggest reducing max_section_length down to 400. Based on my estimates, each Chinese character requires at least 1 token and sometimes more.

You would need to clear out the index and re-run the indexer after changing that.

**Reply by AshleyZhao** [reply]
Sorry that we haven't gotten back for awhile,

From earlier this week, we implemented this solution and the results improved. However, we suspect that the text is still not being chunked properly, as the indexer returns an unusually high number of tokens for pdfs in Chinese.

Perhaps the delimiters are a problem:
<img width="683" alt="image" src="https://github.com/microsoft/AI-Chat-App-Hack/assets/22757766/9603d326-15f4-4a48-ad75-8bb60d326a7b">

And also this:
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/22757766/227263cf-55ba-4877-af6c-37c9010f3b93)

We plan to update the textsplitter.py file and hope that is the correct direction, but let us know if there are better suggestions?

**Reply by pamelafox** [reply]
You're right, the splitting isn't correct for CJK languages, we've got an issue discussing that here:
https://github.com/Azure-Samples/azure-search-openai-demo/issues/867

@tonybaloney also recently did a lot of work into researching tokenization for Japanese words, maybe some of his learnings can benefit you. See https://github.com/tonybaloney/azure-openai-japanese-best-practice
